### PR TITLE
feat: Redis key-value capability implementation

### LINKED ISSUE #187: make redis implementor for kv capability
**Describe the solution you'd like**
see title

**Additional context**
n/a

### PR DESCRIPTION
**Pull request by Mossaka**
This PR adds Redis implementation for the key-value capability. The configuration takes in a connection string like `redis://127.0.0.1:6379`. To be able to run the example, Redis server needs to be installed and started. 

image

This will close #187

### CODE REVIEW

**Comment by Mossaka** [comment]
The redis integration test will need to add a bit more logic to ensure that `redis-server` is installed and running properly. I will defer that to another PR.

File: crates/kv/src/implementors/redis.rs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            )
+            .with_context(|| {
+                format!(
+                    "failed to get 'REDIS_ADDRESS' secret using secret store type: {}",
+                    slight_state.secret_store
+                )
+            })
+            .unwrap(),
+        )
+        .unwrap();
```
Comment (danbugs) [inline crates/kv/src/implementors/redis.rs:33]: could you refactor this to use `get_from_state(.)` instead?

File: crates/kv/src/implementors/redis.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        Self { client }
+    }
+
+    pub fn get(&self, key: &str) -> Result<Vec<u8>> {
+        let mut con = self.client.get_connection()?;
+        let val: Vec<u8> = con.get(key)?;
+        // Redis GET returns [:ok; nil] for non-existent keys
+        if val.is_empty() {
+            return Err(anyhow::anyhow!("key not found"));
+        }
```
Comment (danbugs) [inline crates/kv/src/implementors/redis.rs:45]: just checking, do you need to close the connection here?

File: crates/kv/src/implementors/redis.rs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        let client = redis::Client::open(connection_string).unwrap();
+        Self { client }
+    }
+
+    pub fn get(&self, key: &str) -> Result<Vec<u8>> {
+        let mut con = self.client.get_connection()?;
+        let val: Vec<u8> = con.get(key)?;
+        // Redis GET returns [:ok; nil] for non-existent keys
+        if val.is_empty() {
+            return Err(anyhow::anyhow!("key not found"));
```
Comment (danbugs) [inline crates/kv/src/implementors/redis.rs:44]: use `bail!` instead of `return Err(anyhow::anyhow!(.))`

**Review by danbugs — CHANGES_REQUESTED** [review]
LGTM~ just a couple of small changes requested!

File: crates/kv/src/implementors/redis.rs (hunk 1 above)
Comment (Mossaka) [reply crates/kv/src/implementors/redis.rs:33]: sure

File: crates/kv/src/implementors/redis.rs (hunk 2 above)
Comment (Mossaka) [reply crates/kv/src/implementors/redis.rs:45]: No, the connection will be dropped once the client is out of scope. https://docs.rs/redis/latest/redis/

**Review by danbugs — APPROVED** [review]
LGTM

**Pull request merged**